File: frontend/src/api/solver.ts
// 排课引擎 API:pre-flight 检查、软约束设置、自动排课任务与进度。

import { apiGet, apiPost, apiPut } from '@/api/client'

export interface PreflightIssue {
  level: 'error' | 'warning'
  code: string
  message: string
  subject_type: string
  subject_id: number
}
export interface PreflightReport {
  semester_id: number
  semester_label: string
  ok: boolean
  error_count: number
  warning_count: number
  issues: PreflightIssue[]
  class_count: number
  teacher_count: number
  assignment_count: number
  total_periods: number
}

export interface SoftScore {
  code: string
  name: string
  weight: number
  opportunities: number
  satisfied: number
  violations: number
  penalty: number
  rate: number
  details: string[]
}
export interface SoftReport {
  total_penalty: number
  items: SoftScore[]
}

export type JobStatus = 'queued' | 'running' | 'finished' | 'failed' | 'cancelled'
export type JobPhase = 'solving' | 'explaining'

/** 无解时的一条原因:message 说发生什么事,suggestion 说可以怎么办。 */
export interface ConflictCause {
  code: string
  scope_type: string
  scope_id: number
  scope_name: string
  message: string
  suggestion: string
  relaxable: boolean
  detail: Record<string, number | string>
}
export interface ConflictReport {
  status: string
  source: 'preflight' | 'analysis' | 'none'
  mode: 'each' | 'joint' | 'structural' | ''
  headline: string
  complete: boolean
  relaxable_codes: string[]
  causes: ConflictCause[]
}

export interface UnscheduledCourse {
  // 一项 = 一个排课单位(走班群组含多项成员教学任务),未排节数只算一次
  assignment_ids: number[]
  subject_name: string
  class_names: string[]
  periods: number
  reason: string  // 完全排不下的原因;solver 自行取舍掉的则为空字符串
}

export interface RelaxableOption {
  code: string
  name: string
}

export interface SolveJob {
  job_id: string
  status: JobStatus
  semester_id: number
  source_timetable_id: number
  source_name: string
  max_seconds: number
  elapsed: number
  solutions: number
  objective: number | null
  result_timetable_id: number | null
  result_name: string | null
  error: string | null
  report: SoftReport | null
  phase: JobPhase
  partial: boolean
  conflict: ConflictReport | null
  unscheduled: UnscheduledCourse[] | null
}

export interface ConstraintConfig {
  semester_id: number
  daily_subject_cap: number
  teacher_daily_max: number
  teacher_consecutive_max: number
  weights: Record<string, number>
  weight_names: Record<string, string>
}

export const preflight = (semesterId: number): Promise<PreflightReport> =>
  apiGet(`/solver/preflight?semester_id=${semesterId}`)

export const getConstraintConfig = (semesterId: number): Promise<ConstraintConfig> =>
  apiGet(`/solver/config?semester_id=${semesterId}`)

export const saveConstraintConfig = (
  semesterId: number,
  body: Omit<ConstraintConfig, 'semester_id' | 'weight_names'>,
): Promise<ConstraintConfig> => apiPut(`/solver/config?semester_id=${semesterId}`, body)

export const listRelaxable = (): Promise<RelaxableOption[]> => apiGet('/solver/relaxable')

export const startAutoSchedule = (
  timetableId: number,
  maxSeconds: number,
  options: { allowPartial?: boolean; relax?: string[] } = {},
): Promise<{ job_id: string }> =>
  apiPost(`/timetables/${timetableId}/auto-schedule`, {
    max_seconds: maxSeconds,
    seed: 0,
    allow_partial: options.allowPartial ?? false,
    relax: options.relax ?? [],
  })

export const getSolveJob = (jobId: string): Promise<SolveJob> => apiGet(`/solver/jobs/${jobId}`)
export const stopSolveJob = (jobId: string): Promise<SolveJob> =>
  apiPost(`/solver/jobs/${jobId}/stop`)
export const cancelSolveJob = (jobId: string): Promise<SolveJob> =>
  apiPost(`/solver/jobs/${jobId}/cancel`)

File: backend/app/schemas/solver.py
"""排课引擎相关 schema(pre-flight 报告、软约束设置与达成度)。"""

from pydantic import BaseModel, Field


class PreflightIssue(BaseModel):
    level: str  # error / warning
    code: str
    message: str
    subject_type: str  # teacher / class / room / assignment / semester
    subject_id: int
    detail: dict = {}


class PreflightOut(BaseModel):
    """排课前置检查报告。ok=False 时 errors 必然非空,自动排课应被拦截。"""

    semester_id: int
    semester_label: str
    ok: bool
    error_count: int
    warning_count: int
    issues: list[PreflightIssue] = []
    # 供 UI 显示规模
    class_count: int
    teacher_count: int
    assignment_count: int
    total_periods: int


# ── 软约束设置与达成度(M3-3)────────────
class ConstraintConfigIn(BaseModel):
    """权重 0 = 关闭该项软约束。设置 UI 于 v2 才做,先以 API 调整。"""

    daily_subject_cap: int = Field(default=2, ge=1, le=8)
    teacher_daily_max: int = Field(default=6, ge=1, le=12)
    teacher_consecutive_max: int = Field(default=3, ge=1, le=12)
    weights: dict[str, int] = {}


class ConstraintConfigOut(ConstraintConfigIn):
    semester_id: int
    weight_names: dict[str, str] = {}  # S1 → 「教师偏好时段」


class SoftScoreOut(BaseModel):
    code: str
    name: str
    weight: int
    opportunities: int  # 满分
    satisfied: int      # 得分
    violations: int
    penalty: int
    rate: float
    details: list[str] = []


class SoftReportOut(BaseModel):
    total_penalty: int
    items: list[SoftScoreOut] = []


# ── 冲突定位与部分排课(M3-5)──────────
class ConflictCauseOut(BaseModel):
    code: str  # H3 / H4 / H9 / H10 / structural,或 pre-flight 检查代码
    scope_type: str
    scope_id: int
    scope_name: str
    message: str  # 易懂说明 + 具体数字
    suggestion: str
    relaxable: bool = False
    detail: dict = {}


class ConflictReportOut(BaseModel):
    status: str
    source: str  # preflight / analysis / none
    mode: str  # each / joint / structural
    headline: str
    complete: bool = True
    relaxable_codes: list[str] = []
    causes: list[ConflictCauseOut] = []


class RelaxableOption(BaseModel):
    code: str
    name: str


class UnscheduledCourseOut(BaseModel):
    # 一项 = 一个排课单位(走班群组含多项成员教学任务;未排节数只算一次,见 M6-3)
    assignment_ids: list[int] = []
    subject_name: str
    class_names: list[str] = []
    periods: int
    reason: str = ""  # 完全排不下的原因;solver 自行取舍掉的则为空


# ── 自动排课任务(M3-4)────────────────
class AutoScheduleRequest(BaseModel):
    """timeout 默认 10 分钟(architecture.md §3.3),可设置。"""

    max_seconds: int = Field(default=600, ge=10, le=3600)
    seed: int = Field(default=0, ge=0)
    # 部分排课:允许少数教学任务未排入,并可勾选放宽的硬约束(M3-5)
    allow_partial: bool = False
    relax: list[str] = []


class AutoScheduleAccepted(BaseModel):
    job_id: str


class SolveJobOut(BaseModel):
    job_id: str
    status: str  # queued / running / finished / failed / cancelled
    semester_id: int
    source_timetable_id: int
    source_name: str
    max_seconds: float
    elapsed: float
    solutions: int
    objective: float | None = None
    result_timetable_id: int | None = None
    result_name: str | None = None
    error: str | None = None
    report: SoftReportOut | None = None
    phase: str = "solving"  # solving / explaining
    partial: bool = False
    conflict: ConflictReportOut | None = None
    unscheduled: list[UnscheduledCourseOut] | None = None
